### DISCUSSION TITLE
SuperActorDatabase

### DISCUSSION THREAD

**Discussion by CodyBerenson** [CodyBerenson/PGMA-Modernized#302] [General]
Hello folks!  Jason has revised the NFO Importer that is packaged as part of PGMA.  The agent will not only upload the metadata from your NFO files for a film, it will also match the actors listed in your NFO files against IAFD and return their photo if it finds a match.  

The previous agent, GAnfoMoviesImporter, has been retired.  The replacement agent is NFOImporter.  For further information about how to use and create NFO files, see the Usage Guide section  entitled GAnfoMoviesImporter.  
Cheers!

**Comment by fourstix** [comment]
I currently use the GAnfoMoviesImporter agent for several films that aren't in GEVI, or any other database.  Does NFO support the previous format?  Can you briefly describe any changes that need to be done for the new agent?

**Comment by CodyBerenson** [comment]
Works exactly the same.  No change on the user side, the agent fixes the previously broken IAFD Actor lookup functionality and adopts the code-base of the rest of the PGMA agents.  If you provide an actor in the actor tag, the agent will attempt to match the actor with IAFD and return a headshot...i.e.  

![image](https://github.com/CodyBerenson/PGMA-Modernized/assets/40373801/034749fc-2ae5-46f3-9640-50a801537f8f)

Usage guide has been updated.

**Reply by fourstix** [reply]
Hi, Thank you.   That's great.  I'll test this and make sure.   

One thing I use this agent for is to locally host images for actors where IAFD is incorrect or unavailable.  For example the name "Viper" will show a head shot of a female actress in IAFD, rather than the actor from Latino Fan Club, because that is what matches first.  In addition, several actors from the Golden Age of Porn are not available in IAFD, but it's easy to find an image online.   

Hopefully, the local actor results will over-ride IAFD rather than the other way around.

**Comment by fourstix** [comment]
I stopped Plex, deleted the old agents and copied over the new version of the Agents I use, including the new NFOImporter.bundle.  

I restarted Plex but when I go to configure the NFO Agent, I don't see the NFO Importer agent agent under the GayAdult, GayAdultFilm or GayAdultScene agent in settings.  

What am I missing?

**Comment by fourstix** [comment]
Nevermind.  Second stop  and restart caused the NFO Importer Agent to appear.

**Comment by JPH71** [comment]
That is effing strange..
How many times did you stop and restart plex...
This is an issue I have not come across...

**Comment by JPH71** [comment]
Phew...

**Comment by CodyBerenson** [comment]
Not sure basic plex functionality works as you think it does.  Here's my understanding, verifiable by reviewing the underlying plex database:

The first time John Doe is scraped as an actor on any film/scene in any library, new rows are added to multiple database tables:  a new row for John Doe is added in the **tags** table (specifically, John Doe is value in the **tags.tag** field) in the database and the URL address scraped for John Doe is populated in the **tags.user_thumb_url field**, along with an entry in the **taggings** table that links the actor from the **tags** table to the film/scene from the **metadata_items** table.  The URL is then updated in one of two ways:  1) Each and every time John Doe appears in any future film/scene (new scrape) or 2) a film that he appears in is clicked on in a plex library (which triggers a metadata refresh), the tags.user_thumb_url is updated with the URL from the latest metadata from that specific film.  So, if you're adding a new film/scene that John Doe appears in to any library and using one of the standard PGMA agents (e.g., GEVI, Waybig), the URL will be updated from IAFD, even if at some point earlier you currated John Doe's image via a NFO. I've read this in Plex forums and I've verified this by watching the database tables.  

This is why actor's images update constantly in plex.  When a film/scene is accessed, it does a refresh of the metadata including actor images.  

Sorry to say this, but the only true way to currate metadata including local actor's headshot URLs is by not using scrapers, and doing it manually via NFO.  But, if you're saying that you have a local set of actors's headshots that you use via NFO, the new NFOImporter will use your local headshots if you manually have populated its address in the thumb tags in the NFO.  The challenge only comes for you if that actor appears in any films that may be added or refreshed in your library by agents other than NFO downstream.

Hope this makes sense.

**Reply by fourstix** [reply]
Right.  I get what you're saying, but it's really not too hard for cases like "Viper" where he only appears in Latino fan club films.   All of his films have nfo's.   Other actors that don't exist in IAFD won't have head shots so they can be defined locally.

However the critical thing is that the hosted image needs to come before the IAFD search in the NFO agent, or maybe just have an option to disable the IAFD search in the NFO agent.

Unfortunately , with the new NFO agent I now see this:
<img width="646" alt="Cops and Roughnecks Actors 2" src="https://github.com/CodyBerenson/PGMA-Modernized/assets/5274281/30d33b00-5b4b-46ea-b72b-88613d1c455c">

Where before with the old GAnfoAgent I saw this, because the IAFD search was broken, and so the local image would be used.
<img width="683" alt="Cops and Roughnecks Actors" src="https://github.com/CodyBerenson/PGMA-Modernized/assets/5274281/772cfd1c-186e-4c0d-b66d-65bc0a6d9369">

Do you want me to open an issue for this?  I can grab the logs for this if you need them.

**Reply by CodyBerenson** [reply]
What are your tags for Viper in your NFO?  Please post your NFO.

**Reply by CodyBerenson** [reply]
And, what is the sort order for your Agents?

**Reply by fourstix** [reply]
> What are your tags for Viper in your NFO? Please post your NFO.

Hi, Here's the nfo file I'm using.  GitHub requires the .nfo file in a zip file to be an attachment.

[(Latino Fan Club) - Cops and Ruffnecks.zip](https://github.com/user-attachments/files/15976906/Latino.Fan.Club.-.Cops.and.Ruffnecks.zip)

**Reply by CodyBerenson** [reply]
Thanks for the NFO.  I'm confused, TMM creates both Thumb and Profile tags, but the Profile tags aren't properly formed and the thumb tags are missing.  Do you hand edit the NFO after TMM creates it...or redact to send here?  There are no references to the actor image URLs.  Do you use the TMM setting "Download actor images to .actors" and your folder contains a hidden folder called .actors?  How are you pointing TMM to your image server?

**Reply by fourstix** [reply]
I don't use the Tiny Media Manager, that's what I get when I googled "TMM" for Plex.  Instead I created the nfo files by hand, using information from various web sites and DVD stores.  The profile tag is an emtpy tag.  That's legal XML syntax for a single empty tag.

I host my own images on a web server running under QNAP.  I set the actor location to global and set the location to the web address of the server.

**Reply by CodyBerenson** [reply]
Actually, your XML was originally created by Tiny Media Manager.  Says so, right at the top.  Nothing wrong with doing it by hand, but you're missing a key tag (thumb) that will solve your problems:
   
![image](https://github.com/CodyBerenson/PGMA-Modernized/assets/40373801/2349c2a3-74bf-40b4-b4e6-93f351cda5dd)

**Reply by fourstix** [reply]
Hi Cody,
Thanks.  I believe I used the sample XML file that was with the agent so it probably was created originally with Tiny Media Manager.  I did use the thumb tag originally.

For some reason, the Plex client on my Amazon FireStick did not resolve an internal Lan IP address for my web server (192.xxx.xxx.xxx) correctly.  

But it would resolve an external IP address correctly.  So I used external address in the thumb tag for awhile, but Google Fiber changes my external IP address occasionally, and all the thumb tags would break.   It was a pain to update all the thumb tags in the 70 or so nfo files I have for films not found by the other agents.

So I moved the URL to the global actors field in the agent and deleted all the thumb tags.  Now I only have to update one address with the IP moves.  

I do need to test the internal IP address in the thumb tag again.  They've updated the Plex client for the FireStick many times in the last year or so, and maybe this issue has been fixed.

Sorry for the long response.

**Comment by CodyBerenson** [comment]
To illustrate, Harold Lopez is an actor that appears in multiple films and scenes across multiple libraries that I have.  In the tags table, there is just one entry that gets updated on every new scrape or refresh where Harold Lopez is an actor:

![image](https://github.com/CodyBerenson/PGMA-Modernized/assets/40373801/efa63bbd-a99a-4987-91cc-0094c4c9b641)

And here are all of the films/scenes that Harold Lopez appears in across all libraries:
![image](https://github.com/CodyBerenson/PGMA-Modernized/assets/40373801/41d3be3c-dd4f-4826-b913-17f1523958d6)

The value in the tags URL field comes from the metadata from the very last film/scene that was added or refreshed.

**Comment by fourstix** [comment]
Yes, but Harold Lopez is recent and has a picture in IAFD.  Try it again with the movie I listed: Cops and Ruffnecks  :-)
Since none of the actors are found.  There's no pictures and no description and no poster.  Nothing, nada.

That's a case when previously I would use the GANFO agent to define things, and now I'd like to use the NFO Importer agent.

**Comment by fourstix** [comment]
Good point on the sort order.  It shows up on the bottom of the list.  I'll put it at the top, since if I have an .nfo file I'd like it to override everything else.  (All the other agents usually don't match anyway, so this may not fix anything.)  I think the problem is the NFO agent itself prioritizes IAFD searches above the local ones.

**Reply by CodyBerenson** [reply]
From my testing, the agent ignores searching IAFD actors who have the **thumb** metadata populated

**Comment by CodyBerenson** [comment]
You have the option of using the Plex standard NFO importer.  That may be your best solution.  We do want actors pictures pulled from IAFD as the go-to.

**Comment by fourstix** [comment]
Ah, maybe I could just add <thumb> <thumb> to all the actor tags, but that would be a big pain for me as a have several of these.  Maybe 30 or so films defined by .nfo.   

But since I specified a global actors folder in the options, it would be best if NFO tried that first, and then if there's no match used IAFD.
![GA NFO Agent Config 3](https://github.com/CodyBerenson/PGMA-Modernized/assets/5274281/45af8d37-fa2a-4910-b1f4-df0a8565bc1c)

An option to use or not use IAFD would also do the trick.

**Comment by fourstix** [comment]
I'm not aware of a standard Plex NFO importer, are you referring to the original agent that was replaced by the NFO agent?

**Reply by CodyBerenson** [reply]
Here is the standard importer for movies:  https://github.com/gboudreau/XBMCnfoMoviesImporter.bundle

Here is the standard importer for tv shows:  https://github.com/gboudreau/XBMCnfoTVImporter.bundle

The PGMA NFO Importer was based on this agent, with IAFD functionality woven in, which broke, and is now fixed.

**Comment by fourstix** [comment]
I tried the XBMCnfo importer, long ago, and I found that wasn't a very good fit for these movies.   The PGMA NFO agent worked better, even with the broken IAFD seach.  

The thumbs is an option, but when the web server IP address changes, then one has to edit every actor entry in every nfo file. Which is quite a pain.   I don't want to set up Dynamic DNS for this one tiny web server that I maintain.

If you really don't want to have an option or modificaion, I think I will modify the PGMA NFO Importer so that it meets my needs better.  I can fork the code and just move the IAFD logic after the global actors logic.  

To me, the agent in its current format isn't very usable since it adds very little of the other agents.  (The only good use case I had for it no longer works.)

**Reply by CodyBerenson** [reply]
appreciate your feedback, sounds like a plan.  take care.

**Comment by JPH71** [comment]
So in regards to this nfo importer do we need an option to ignore IAFD

**Reply by fourstix** [reply]
Yes, please!  :-)
I think that would improve the function.  I would be glad to help anyway I can.  If you want I can send you information to help test this with an actual image server.

**Comment by fourstix** [comment]
PS.  I use this agent with Triga films.  :-)

**Comment by JPH71** [comment]
The image server thingy would help a hell of a lot..
We can program something so that the image server is checked first before
IAFD

This will mean that once you have picked an image, it will be used as the
default irrespective of whether there is an image or not on iafd..

We can use the image address on iafd as the key in a database.

We could also create an online database which could be updated once each
time an image is not found on iafd and a user has supplied one.

This would solve the actors and directors collection issue as cloudflare is
preventing the download of their images..

This image server is interesting as it would reduce hitting IAFD with
multiple requests and improve speed... as its local...

Jason xxx

**Reply by CodyBerenson** [reply]
Sounds good.  Who's creating?  Who's hosting?  Who's maintaining?  Here's who:  IAFD.  lmao.

**Reply by fourstix** [reply]
It's not that hard to host your own.  On QNAP I host them on a web server, and then just supply the head shots in a directory for the actors that aren't available on IAFD in my own set of films.   

The web server on my NAS box is not set up for a world wide traffic, but works just fine for my own Plex server.

**Comment by CodyBerenson** [comment]
@fourstix and @JPH71:

None of any of this has anything to do with the new agent.  Please open a new enhancement request for either the IAFD checking flag on/off  and/or SuperActorDatabase.  I'm going to close this thread and reopen a new one to announce the new Agent.  Thanks.

**Comment by JPH71** [comment]
Yes, this should be an enhancement...
If I can work out how to access locally stored images, this would solve
multiple issues...

**Comment by fourstix** [comment]
Holy crap!  I looked at the code and when the actor thumb is set to "global" it should work the way I'm requesting.  

For some reason the agent didn't accept the change from "link" to "global".  I made the change, again, and restarted the Plex Server and this time it picked up the change and the NFO agent will use the global image instead of searching IAFD.

**Comment by JPH71** [comment]
Cool that is sorted... plex can be a right pain in the nethers at times...

But if you can give me an insight to the Web servers let me know...

**Comment by fourstix** [comment]
Be glad to...    I sent you an email.   It's dead simple in Qnap.  

https://docs.qnap.com/operating-system/qts/4.5.x/en-us/GUID-1C3A7404-D511-48A9-995B-6E75C880AD1D.html

Basically one enables the service, and puts the web contents underneath a predefined directory named /Web.  Any path to the file would be appended to the base web service URL.

**Comment by jfreiman** [comment]
Just like the official Plex metadata scraper as it regards to "Cast" members, there are many actors with the same name.

Plex's own agents a way to deal with this, would it be possible to use each cast member's IAFD ID [Grayson Frost] (https://www.iafd.com/person.rme/id=08073200-56d8-45e9-b542-1ac5ffb0f9e6l) found in the URL (id=08073200-56d8-45e9-b542-1ac5ffb0f9e6) to specifically specify a particular cast member?

This would prevent PGMA/Plex from replaceing existing cast members with the images and search function from being overwritten by the information of the last cast member added,,,,

**Comment by JPH71** [comment]
This would be possible only with an nfo file.

Jason
